scimax-org-babel-ipython.el

```diff
@@ -1134,16 +1134,20 @@ This function is used in a C-c C-c hook to make it work like other org src block
   "Execute all the ipython blocks in the buffer up to point."
   (interactive)
   (let ((s (org-babel-get-session))
-        (p (point)))
+        (l (- (point-max) (point))))
     (save-excursion
       (goto-char (point-min))
       (while (and (org-babel-next-src-block)
-                  (<= (point) p))
-        (and (string= (first (org-babel-get-src-block-info)) "ipython")
-             (string= (org-babel-get-session) s)
-             (if org-babel-async-ipython
-                 (org-babel-execute-async:ipython)
-               (org-babel-execute-src-block)))))))
+                  (<= (point) (- (point-max) l)))
+        (when (and (string= (first (org-babel-get-src-block-info)) "ipython")
+                   (string= (org-babel-get-session) s))
+          (if org-babel-async-ipython
+              (progn
+                ;; wait until last cell is finished
+                (while (ob-ipython-get-running)
```

Comment:
why is this the right thing to do? Won;t this effectively block emacs on each cell? Shouldn't this just build up the queue to be run?

Reply:
> Shouldn't this just build up the queue to be run?

I don't know why but that isn't the case. With the cell running, the queue won't run and leave a red queue link in the results.

Reply:
You can reproduce the issue by following steps:

```org
#+begin_src ipython
(print 1)
#+end_src

#+begin_src ipython
(print 2)
#+end_src

#+begin_src ipython
(print 3)
#+end_src

#+begin_src ipython
(print 4)
#+end_src
```
Put the cursor on src block 4, call the function and you'll see what I mean.

Reply:
Actually, this should not be called a fix. It is my workaround on the issue. So if you can fix it, then we don't need this. But I'm afraid I can't help on that.

Reply:
No worries, there is definitely some bad behavior in these functions. Thanks for your help in trying to clear them up. It is complicated by the intermingling of async and sync code, and that is probably what should get fixed eventually. The ob-ipython author suggested the kernel has its own queue, which would simplify things alot, but I haven't figured out how to get the results back where they belong in an async mode.

In your example of four blocks above there may be a confounding issue with exceptions (it looks like you wrote lisp in the python blocks :). That is another issue I am trying to figure out too.

Reply:
Sorry for the typos! I thought I was writing Python! lol. 

I've updated that.

Reply:
Thanks for reviewing. I agree this is a complicated issue. But while you're looking for the real solution, maybe you could merge this as a temporary fix ? I think scimax-org-babel-ipython is one of the main reasons why people choose scimax and those functions are used often. So a temporary fix may not be a bad thing as it looks.